### PR TITLE
Post Comments Form: Add button that enables commenting to warning

### PR DESCRIPTION
**Pull request by ockham**
## What?
Add a button to the Post Comments Form block that allows the user to enable commenting when it's disabled.
Part of #40614; also the first experiment to explore whether a solution like this makes sense.

## Why?
Remove frustration from the user coming from a warning message that doesn't give them agency to remediate the problem. See #40614 for more context.

## How?
In the Post Comments Form block warning that states that commenting is not allowed, add a button that allows the user to enable it.

## Testing Instructions
- Create a new post.
- Open the inspector sidebar for the Post.
- Disable comments for the post. 
image
- Add the Comments block to the post.
- Verify that the warning that states that comments aren't allowed includes a button to enable them.
- Click that button and verify that the warning disappears, and the comment form appears.

_Note that this might not be the most interesting case from the list in #40614, since the "Allow comments" option is otherwise fairly easily accessible from the Post sidebar. However, we might apply the same principle (button in warning) to other cases where the corresponding setting isn't as readily availably (but e.g. somewhere in wp-admin > Settings > Discussion, see e.g. #41308)._

## Screenshots or screencast

enable-comments

## Open questions

- Do we need to limit this based on user capabilities? (It seems like the [option in the Post sidebar](https://github.com/WordPress/gutenberg/blob/5e09117a2bab1f08e529aa9c0111f31efe0aca6f/packages/editor/src/components/post-comments/index.js) isn't really restricted based on caps, so maybe they're available to anyone with editing capabilities.)
- Separations of concerns, see https://github.com/WordPress/gutenberg/pull/41603#discussion_r892728680.

### CODE REVIEW

File: packages/block-library/package.json
Comment (ockham) [inline packages/block-library/package.json]: This is probably the most controversial change in this PR 😬  In the past, we've been keen to avoid having the blocks depend on the editor.

The reason we need this is the `editPost` action we're calling [here](https://github.com/WordPress/gutenberg/pull/41603/files#diff-abf88ae09f89f6af7894b1f8480350eb820df74aff7e2b61490e93dd48c4722fR74-R77).

I would argue that this make sense, and that we might need access to other "editor" settings if we implement other items from the list in https://github.com/WordPress/gutenberg/issues/40614.

The functionality of blocks like the Post Comments Form or the Avatar (see e.g. #41308) are directly determined by these settings, so IMO, we're not unduly mixing concerns. Instead, I think that we'll find that a growing number of "theme" blocks rely on some settings that up to now we've thought of as "editor" settings (but that are really post-specific or site-wide).

If we're still adamant that the block library package shouldn't depend on `@wordpress/editor`, there might be some strategies to mitigate:

- Move relevant selectors and actions (such as `editPost`) to a different package (that better represents what they are, e.g. `@wordpress/site` or `@wordpress/post` (or `-data`)).
- Move theme blocks (that rely on those selectors and/or actions) to a different package (`@wordpress/theme-blocks`)

Curious to hear thoughts from @mcsf @gziolo et al :)

Comment (mcsf) [reply packages/block-library/package.json]: > strategies to mitigate:
> 
> * Move relevant selectors and actions (such as `editPost`) to a different package (that better represents what they are, e.g. `@wordpress/site` or `@wordpress/post` (or `-data`)).
> * Move theme blocks (that rely on those selectors and/or actions) to a different package (`@wordpress/theme-blocks`)

Between the two, `theme-blocks` seems more likely, as that is compatible with the idea that the `block-editor` and `block-library` modules should remain as backend-agnostic as feasible. That said, have you pondered alternatives like loose coupling via dynamic typing?

```js
const editorDispatcher = useDispatch( 'core/editor' ); // note passing of string literal
...
if ( editorDispatcher ) { // note gating of comment-enabling feature
  const enableComments = () => editorDispatcher.editPost( { .... } );
  actions = [ ... ];
}
```

(Dynamic imports could also be added to the mix if we care about calling `useDispatch` with the actual imported `editorStore`.)

This is not something to which I've given enough thought, just a low-commitment idea that seems worth debating. I would say that, conceptually, the single thing that I like about it is that it carries with it the spirit of progressive enhancement / graceful degradation — something we tend to recommend when using Web APIs (e.g. `if ( window.MutationObserver ) ...`).

Comment (ockham) [reply packages/block-library/package.json]: Thanks @mcsf, I hadn't really thought of loose coupling to sidestep the problem. Trying it on for size: d1a9f9717981dfd08429b9bd6d8f02252d2ae94a

I think this is viable in the short term. I'm not sure if it's the be-all, end-all in the long term. We might see a growing number of FSE blocks that need information about the post or site (the "Are Avatars enabled in discussion settings?" being another example), which have so far been considered _editor_ settings, but which we might want to revisit.

(I'm happy to move forward with the loose coupling for now, just wanted to leave this comment for visibility, since I'm referring to this discussion from the code 🙂)

**Review by cbravobernal — APPROVED** [review]
Nice job! 
It is great to see it work, but, I guess this case is not really common. 
As you have to:
1.) Add the comment form to a post. While the common case is to add it on the site editor.
2.) Have global comments enabled.
3.) Disable the comments in each post after adding the comment.

However, dealing with global settings as you mentioned could have huge benefits on the site editor part (as a post edition should not be able to change global settings, this could mess the site with side effects).

Should we explore that possibility for other use cases like Comments, Avatar, Date formatting, etc?

**Comment by ockham** [comment]
> Nice job! 
> It is great to see it work, but, I guess this case is not really common. 
> As you have to:
> 1.) Add the comment form to a post. While the common case is to add it on the site editor.
> 2.) Have global comments enabled.
> 3.) Disable the comments in each post after adding the comment.

Yeah -- plus the post settings are easily available from the sidebar, so it's almost redundant 😅 

This is really mostly to demo the principle, since it was one of the easiest [settings to change via code](https://github.com/WordPress/gutenberg/pull/41603/files#diff-abf88ae09f89f6af7894b1f8480350eb820df74aff7e2b61490e93dd48c4722fR82-R85).

> However, dealing with global settings as you mentioned could have huge benefits on the site editor part (as a post edition should not be able to change global settings, this could mess the site with side effects).

👍 

> Should we explore that possibility for other use cases like Comments, Avatar, Date formatting, etc?

I would like to try that, yes! Issue https://github.com/WordPress/gutenberg/issues/40614 has a whole list of other warnings; I'd definitely like to prioritize the ones that occur in the Site Editor.

I'm just not sure if we even have actions available for dispatch for all of these 😅 (or even `POST` endpoints, in some cases) -- this might be the bulk of the work for some of those.

File: packages/block-library/package.json
Comment (gziolo) [reply packages/block-library/package.json]: The same issue is being tracked at https://github.com/WordPress/gutenberg/issues/32842. I believe that the best way to move forward is to move all the handling of the post from the `editor` store to entities in `@wordpress/core-data`. I believe it should be possible as of today, but I haven't tried.

Comment (mcsf) [reply packages/block-library/package.json]: > The same issue is being tracked at #32842

Thanks for finding that one.

> move all the handling of the post from the `editor` store to entities in `@wordpress/core-data`. I believe it should be possible as of today

Yes, this is my suspicion too, but I don't know the entity-based APIs well enough to advise.

Comment (mcsf) [reply packages/block-library/package.json]:
```js
const MY_TYPE : 'post' | 'page' | ... = ...
const MY_ID : number = ...

// reading
const { comment_status } = useSelect( 'core' ).getEntityRecord( 'postType', MY_TYPE, MY_ID );
// writing
useDispatch( 'core' ).editEntityRecord( 'postType', MY_TYPE, MY_ID, { comment_status: 'open' } );
// saving
await useDispatch( 'core' ).saveEditedEntityRecord( 'postType', MY_TYPE, MY_ID );
```

This is the standard way approach for manipulating entities, as far as I'm aware. The only awkward bit that I see at the moment is that saving an edited entity seems to be all or nothing, so you may want to **not** call `saveEditedEntityRecord` and let the user manually save the whole post at the end when they are done — which is maybe a better UX anyway, no?

(Also, thanks to @adamziel for his [Core Data guides](https://developer.wordpress.org/block-editor/how-to-guides/data-basics/3-building-an-edit-form/) and for making it easier with types to find the expected shape of [Post](https://github.com/WordPress/gutenberg/blob/65550e38a68befd30250a05a05d76a7d42ebfc22/packages/core-data/src/entity-types/post.ts#L111-L115) and [CommentStatus](https://github.com/WordPress/gutenberg/blob/65550e38a68befd30250a05a05d76a7d42ebfc22/packages/core-data/src/entity-types/helpers.ts#L22).)

Comment (ockham) [reply packages/block-library/package.json]: Thanks @gziolo, #32842 is an interesting read.

> move all the handling of the post from the `editor` store to entities in `@wordpress/core-data`. I believe it should be possible as of today

Not totally sure I'm following. So for this PR, I'm using `editPost` from the editor store. This is defined here: https://github.com/WordPress/gutenberg/blob/65550e38a68befd30250a05a05d76a7d42ebfc22/packages/editor/src/store/actions.js#L130-L144

It should be easy enough to import `editEntityRecord` directly from `core-data`.

OTOH, `getCurrentPost` is defined in the `editor` store and directly relies on that store's state (`state.postId` and `state.postType`) -- which indeed seems to be fairly editor-specific. I'm not sure what we should do about those -- move them into `core-data`'s state?

**Comment by ockham** [comment]
I'll hold off a bit longer before I merge as there's been a bit of a new development on this discussion: https://github.com/WordPress/gutenberg/pull/41603#discussion_r896544446. Maybe we can avoid the (loose) coupling with the editor store after all 😬

File: packages/block-library/package.json
Comment (gziolo) [reply packages/block-library/package.json]: > `getCurrentPost` is defined in the `editor` store and directly relies on that store's state (`state.postId` and `state.postType`) -- which indeed seems to be fairly editor-specific. I'm not sure what we should do about those -- move them into `core-data`'s state?

I examined the `editor` store more closely. It looks like all handling for post/page/CPT is now fully controlled by `@wordpress/core-data`. The only part where `editor` store is helpful, as pointed out by @ockham, is when it automatically infers the post id and type. This is something that can be set through the [Block Context](https://github.com/WordPress/gutenberg/blob/trunk/docs/reference-guides/block-api/block-context.md). It looks like the Post Comments Form block has access to both post id and type:

https://github.com/WordPress/gutenberg/blob/d1a9f9717981dfd08429b9bd6d8f02252d2ae94a/packages/block-library/src/post-comments-form/block.json#L14

so this should work perfectly fine.

Comment (mcsf) [reply packages/block-library/package.json]: The block context, as reminded by gziolo, is definitely your friend.

I wanted to get a personal feel for these APIs, so I got this working in my environment, preferring `useEntityProp` over the other functions that I previously mentioned:

```js
import { useEntityProp } from '@wordpress/core-data';

function MyBlockEdit( { context } ) {
  const { postId, postType } = context;
  const [ commentStatus, setCommentStatus ] = useEntityProp(
    'postType',
    postType,
    'comment_status'
  );
}
```

_edit: this assumes that the block's metadata includes `"usesContext": [ "postId", "postType" ],`_

Comment (adamziel) [reply packages/block-library/package.json]: Also surfacing `useEntityRecord`:

```js
import { useEntityRecord } from '@wordpress/core-data';

function PageTitleDisplay( { id } ) {
  const { record, isResolving } = useEntityRecord( 'postType', 'page', id );

  if ( isResolving ) {
    return 'Loading...';
  }

  return record.title;
}
```

Comment (adamziel) [reply packages/block-library/package.json]: > The only awkward bit that I see at the moment is that saving an edited entity seems to be all or nothing, so you may want to not call saveEditedEntityRecord and let the user manually save the whole post at the end when they are done — which is maybe a better UX anyway, no?

Partial save would certainly be a nice feature! I don't we have it today, but you could also pass a custom `__unstableFetch` option and pre-process the http request body before sending it through. It feels hacky, though, so perhaps do that only if it's really important :-)

Comment (mcsf) [reply packages/block-library/package.json]: > Also surfacing `useEntityRecord`:

The issue with `useEntityRecord`, at least in my testing, was that it doesn't consider _edited_ entities. That is, toggling `comment_status` on the post entity doesn't change the entity record until the entity is saved.

While that doesn't make it a dealbreaker for the comment-enabling functionality debated in this PR, it would still require some manual synchronisation between the edited entity, the warning UI, and the saving procedure. That's why `useEntityProp` felt better to me in the end. Or am I missing something with `useEntityRecord`, @adamziel?

Comment (adamziel) [reply packages/block-library/package.json]: Ah no, you're right @mcsf. The support for edits is coming, but isn't there yet: https://github.com/WordPress/gutenberg/pull/39595

Comment (ockham) [reply packages/block-library/package.json]: Thanks a lot, folks! I'm gonna give that (i.e. `useEntityProp` with post ID and type from block context) a try!

Comment (ockham) [reply packages/block-library/package.json]: Well, that was easy: https://github.com/WordPress/gutenberg/pull/41603/commits/dda7509644ee2b694757b4ad7e72afe38b17644c

Turns out we were already using `useEntityProp` and block context to read the comment status anyway 😬

**Comment by ockham** [comment]
Ready for another look 😄

File: packages/block-library/package.json
Comment (mcsf) [reply packages/block-library/package.json]: 😆 wonderful!

File: packages/block-library/src/post-comments-form/edit.js
Comment (mcsf) [inline packages/block-library/src/post-comments-form/edit.js]: Could use a `translators` note, IMO.

File: packages/block-library/src/post-comments-form/edit.js
Comment (mcsf) [inline packages/block-library/src/post-comments-form/edit.js]: I guess this could now be inlined, but it's a matter of style. :)

**Review by mcsf — APPROVED** [review]

File: packages/block-library/src/post-comments-form/edit.js
Comment (ockham) [reply packages/block-library/src/post-comments-form/edit.js]: Good point 😄 b95cc0461be3c22a7fa8ab18a3ed30b533fdc3b1

File: packages/block-library/src/post-comments-form/edit.js
Comment (ockham) [reply packages/block-library/src/post-comments-form/edit.js]: What kind of information is most needed here?

Something like

```suggestion
					{ _x( 'Enable comments', 'button label' ) }
```

or more like

```suggestion
					{ _x( 'Enable comments', 'action that affects the current post' ) }
```

? 😄

Comment (mcsf) [reply packages/block-library/src/post-comments-form/edit.js]: More the latter. :)

Comment (ockham) [reply packages/block-library/src/post-comments-form/edit.js]: d72be81

**Review by mcsf — APPROVED** [review]

**Comment by ockham** [comment]
Trying a rebase to fix e2e failures.

**Comment by ockham** [comment]
The [remaining e2e failures](https://github.com/WordPress/gutenberg/runs/6936252589?check_suite_focus=true) seem unrelated (Navigation block); it looks like they're also present on `trunk` currently.

**Comment by adamziel** [comment]
I just cherry-picked this PR to the wp/6.0 branch to get it included in the next release: 4970dcfafd

**Pull request merged**
